components/user-mgt/org.wso2.carbon.identity.user.pre.update.profile.action/src/main/java/org/wso2/carbon/identity/user/pre/update/profile/action/internal/execution/PreUpdateProfileRequestBuilder.java

```diff
@@ -0,0 +1,327 @@
+import org.wso2.carbon.identity.user.pre.update.profile.action.internal.model.UpdatingUserClaim;
+import org.wso2.carbon.user.api.UserRealm;
+import org.wso2.carbon.user.api.UserStoreException;
+import org.wso2.carbon.user.api.UserStoreManager;
+import org.wso2.carbon.user.core.UniqueIDUserStoreManager;
+import org.wso2.carbon.user.core.UserCoreConstants;
+import org.wso2.carbon.user.core.service.RealmService;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.regex.Pattern;
+
+/**
+ * Pre Update Profile Action Request Builder.
+ */
+public class PreUpdateProfileRequestBuilder implements ActionExecutionRequestBuilder {
+
+    private static final String ROLE_CLAIM_URI = "http://wso2.org/claims/roles";
+    private static final String GROUP_CLAIM_URI = "http://wso2.org/claims/groups";
+
+    @Override
+    public ActionType getSupportedActionType() {
+
+        return ActionType.PRE_UPDATE_PROFILE;
+    }
+
+    @Override
+    public ActionExecutionRequest buildActionExecutionRequest(FlowContext flowContext,
+                                                              ActionExecutionRequestContext actionExecutionContext)
+            throws ActionExecutionRequestBuilderException {
+
+        UserActionContext userActionContext =
+                flowContext.getValue(UserActionContext.USER_ACTION_CONTEXT_REFERENCE_KEY, UserActionContext.class);
+        PreUpdateProfileAction preUpdateProfileAction = (PreUpdateProfileAction) actionExecutionContext.getAction();
+
+        return new ActionExecutionRequest.Builder()
+                .actionType(getSupportedActionType())
+                .event(getEvent(userActionContext, preUpdateProfileAction))
+                .build();
+    }
+
+    private Event getEvent(UserActionContext userActionContext, PreUpdateProfileAction preUpdateProfileAction)
+            throws ActionExecutionRequestBuilderException {
+
+        PreUpdateProfileEvent.Builder eventBuilder = new PreUpdateProfileEvent.Builder();
+        eventBuilder.initiatorType(getInitiatorType());
+        eventBuilder.action(PreUpdateProfileEvent.Action.UPDATE);
+        eventBuilder.request(getPreUpdateProfileRequest(userActionContext));
+        eventBuilder.tenant(getTenant());
+
+        UniqueIDUserStoreManager userStoreManager = getUserStoreManager();
+        eventBuilder.user(getUser(userActionContext, preUpdateProfileAction, userStoreManager));
+        eventBuilder.userStore(getUserStore(userActionContext, userStoreManager));
+
+        return eventBuilder.build();
+    }
+
+    private PreUpdateProfileEvent.FlowInitiatorType getInitiatorType() throws ActionExecutionRequestBuilderException {
+
+        Flow flow = Optional.ofNullable(IdentityContext.getThreadLocalIdentityContext().getFlow())
+                .orElseThrow(() -> new ActionExecutionRequestBuilderException("Unknown flow."));
```

Comment:
Shall we make exception description more descriptive.

Reply:
Well, I think we can go ahead with this as this flow referred is internal and this exception could get logged due to a programming error only, in a situation when a flow is not set in the IdentityContext